clients/src/main/java/org/apache/kafka/clients/consumer/internals/metrics/ConsumerRebalanceMetricsManager.java

```diff
@@ -0,0 +1,110 @@
+    public final MetricName failedRebalanceTotal;
+    public final MetricName failedRebalanceRate;
+    private long lastRebalanceEndMs = -1L;
+    private long lastRebalanceStartMs = -1L;
+
+    @SuppressWarnings("this-escape")
+    public ConsumerRebalanceMetricsManager(Metrics metrics) {
+        super(CONSUMER_METRIC_GROUP_PREFIX + COORDINATOR_METRICS_SUFFIX);
+
+        rebalanceLatencyAvg = createMetric(metrics, "rebalance-latency-avg",
+                "The average time taken for a group to complete a rebalance");
+        rebalanceLatencyMax = createMetric(metrics, "rebalance-latency-max",
+                "The max time taken for a group to complete a rebalance");
+        rebalanceLatencyTotal = createMetric(metrics, "rebalance-latency-total",
+                "The total number of milliseconds spent in rebalances");
+        rebalanceTotal = createMetric(metrics, "rebalance-total",
+                "The total number of rebalance events");
+        rebalanceRatePerHour = createMetric(metrics, "rebalance-rate-per-hour",
+                "The number of rebalance events per hour");
+        failedRebalanceTotal = createMetric(metrics, "failed-rebalance-total",
+                "The total number of failed rebalance events");
+        failedRebalanceRate = createMetric(metrics, "failed-rebalance-rate-per-hour",
+                "The number of failed rebalance events per hour");
+
+        successfulRebalanceSensor = metrics.sensor("rebalance-latency");
+        successfulRebalanceSensor.add(rebalanceLatencyAvg, new Avg());
+        successfulRebalanceSensor.add(rebalanceLatencyMax, new Max());
+        successfulRebalanceSensor.add(rebalanceLatencyTotal, new CumulativeSum());
+        successfulRebalanceSensor.add(rebalanceTotal, new CumulativeCount());
+        successfulRebalanceSensor.add(rebalanceRatePerHour, new Rate(TimeUnit.HOURS, new WindowedCount()));
+
+        failedRebalanceSensor = metrics.sensor("failed-rebalance");
+        failedRebalanceSensor.add(failedRebalanceTotal, new CumulativeSum());
+        failedRebalanceSensor.add(failedRebalanceRate, new Rate(TimeUnit.HOURS, new WindowedCount()));
+
+        Measurable lastRebalance = (config, now) -> {
+            if (lastRebalanceEndMs == -1L)
+                return -1d;
+            else
+                return TimeUnit.SECONDS.convert(now - lastRebalanceEndMs, TimeUnit.MILLISECONDS);
+        };
+        lastRebalanceSecondsAgo = createMetric(metrics,
+                "last-rebalance-seconds-ago",
+                "The number of seconds since the last rebalance event");
+        metrics.addMetric(lastRebalanceSecondsAgo, lastRebalance);
+    }
+
+    public void recordRebalanceStarted(long nowMs) {
+        lastRebalanceStartMs = nowMs;
+    }
+
+    public void recordRebalanceEnded(long nowMs) {
+        lastRebalanceEndMs = nowMs;
+        successfulRebalanceSensor.record(nowMs - lastRebalanceStartMs);
+    }
+
+    public void maybeRecordRebalanceFailed() {
```

Comment:
Should we just go with recordRebalanceFailed()?

Reply:
This PR is refactoring earlier code. This method name is from the earlier class and I don't think it should be renamed as part of the refactoring. It is still doing a conditional record of the metrics.